tools/shell/embedded_shell.cpp

```diff
@@ -886,438 +897,452 @@ void EmbeddedShell::printExecutionResult(QueryResult& queryResult) const {
-            colsWidth[i] = std::max(queryResult.getColumnNames()[i].length(),
-                               queryResult.getColumnDataTypes()[i].toString().length()) +
-                           2;
+        if (queryResult.hasNextWarning()) {
+            printf("Warnings:\n");
+            printTruncatedExecutionResultImpl(FetchQueryWarnings{queryResult});
         }
-        uint64_t rowCount = 0;
-        while (queryResult.hasNext()) {
-            if (numTuples > maxRowSize && rowCount >= (maxRowSize / 2) + (maxRowSize % 2 != 0) &&
-                rowCount < numTuples - maxRowSize / 2) {
-                auto tuple = queryResult.getNext();
-                rowCount++;
-                continue;
-            }
-            auto tuple = queryResult.getNext();
-            for (auto i = 0u; i < colsWidth.size(); i++) {
-                if (tuple->getValue(i)->isNull()) {
-                    continue;
+        const auto printResult = printTruncatedExecutionResultImpl(FetchQueryResults{queryResult});
+
+        // print query result (numFlatTuples & tuples)
+        if (stats) {
+            if (printResult.numTuples == 1) {
+                printf("(1 tuple)\n");
+            } else {
+                printf("(%" PRIu64 " tuples", printResult.numTuples);
+                if (printResult.rowTruncated) {
+                    printf(", %" PRIu64 " shown", maxRowSize);
                 }
-                std::string tupleString = tuple->getValue(i)->toString();
-                uint32_t fieldLen = 0;
-                uint32_t chrIter = 0;
-                while (chrIter < tupleString.length()) {
-                    fieldLen += Utf8Proc::renderWidth(tupleString.c_str(), chrIter);
-                    chrIter =
-                        utf8proc_next_grapheme(tupleString.c_str(), tupleString.length(), chrIter);
+                printf(")\n");
+            }
+            if (printResult.colsWidth.size() == 1) {
+                printf("(1 column)\n");
+            } else {
+                printf("(%" PRIu64 " columns", (uint64_t)printResult.colsWidth.size());
+                if (printResult.colTruncated) {
+                    printf(", %" PRIu64 " shown", printResult.colsPrinted);
                 }
-                // An extra 2 spaces are added for an extra space on either
-                // side of the std::string.
-                colsWidth[i] = std::max(colsWidth[i], fieldLen + 2);
+                printf(")\n");
             }
+            printf("Time: %.2fms (compiling), %.2fms (executing)\n",
+                querySummary->getCompilingTime(), querySummary->getExecutionTime());
+        }
+    }
+}
+
+template<QueryResultFetcher Fetcher>
+EmbeddedShell::PrintResult EmbeddedShell::printTruncatedExecutionResultImpl(
+    Fetcher queryResult) const {
+    auto tableDrawingCharacters =
+        common::ku_dynamic_cast<DrawingCharacters*, BaseTableDrawingCharacters*>(
+            drawingCharacters.get());
```

Comment:
I recommend hiding whitespace in your diff viewer since I unindented a 400+ line function here :)

Reply:
shell changes were reverted